### DISCUSSION TITLE
Ability to create new event bindings within a binding. It's bindings all the way down.

### DISCUSSION THREAD

**Discussion by timacdonald** [junegunn/fzf#2532] [Ideas]
With FZF currently, it is possible to create event bindings. You can bind to events like `change` and `ctrl-r` press and perform certain actions, such as (silly example) changing the prompt...

```
  let spec = {'options': [
      \ '--phony',
      \ '--bind',
      \ 'ctrl-r:change-prompt(R > )'
      \ '--bind',
      \ 'ctrl-f:change-prompt(F > )',
      \ ]
```

This also allows for search setups such as those demonstrated here that rely on listening to `change` event to perform a "live grep" search, which is big 😍 vibes: https://github.com/junegunn/fzf/blob/master/ADVANCED.md#switching-to-fzf-only-search-mode

This is enabled by the ability to _remove_ a event binding within an event binding, i.e. when I press `ctrl-f` remove the `change` binding. However, currently there is no way to _add_ an event binding within a binding.

Having the ability to add an event binding within a binding would allow for flows such as...

```
Open VIM with a live grep search (already possible).

Press `ctrl-r` to switch to a regex based search and keep the live grep search

Press `ctrl-f` to switch the file filtering with FZF based on the search results

Press `ctrl-r` to switch back to the RG regex search and keep the live grep search
```

This is not yet possible because there is no way to bind to the `change` event within a key binding. I imagine the spec would look like the following...

```
  let spec = {'options': [
      \ '--phony',
      \ '--prompt',
      \ 'Plain search: ',

      ' Default binding...
      \ '--bind',
      \ 'change:reload:'.plain_reload_command,

      ' Switch to regex based live search....
      \ '--bind',
      \ 'ctrl-r:unbind(change)+change-prompt(Regex search: )+disable-search+@change:reload:'.regex_reload_command,

      ' Switch to non-regex based live search....
      \ '--bind',
      \ 'ctrl-s:unbind(change)+change-prompt(Plain search: )+disable-search+@change:reload:'.plain_reload_command,

      ' Switch to FZF result filtering....
      \ '--bind',
      \ 'ctrl-f:unbind(change)+change-prompt(Result filter: )+enable-search+clear-query',
      \ ]}
```

Used `@` just to signify that it is an event binding

My current setup has me setting two different key mappings. One for regex searches and one for plain text searches. To change between to the two I need to exit and re-enter the search. The current setup also means that after I switch to the FZF filtering stage, there is no way to return and fix / adjust the initial RG based search.

I think being able to switch between all these at will would be pretty neat. I attempted to PR this, but I've not played with GO before. It someone with more experience with GO or FZF would have any interested in working on something like this together (assuming it is a feature that would be accepted), I'd be keen to help make it happen.

**Comment by vovcacik** [comment]
I also got close to what you describe, but can't get "live" updates without rebinding _change_ event. https://github.com/junegunn/fzf/issues/2465#issuecomment-934522391 with demo.

**Comment by bernardassan** [comment]
I am having a similar limitation with using dynamic rebinding #2752 , maybe we can work today on this as Our first Go issue. I also have little experience with Go but Very welling to learn and help fix this issue
